src/io.rs

```diff
@@ -98,6 +101,56 @@ impl Write for CallbackWriter {
     }
 }
 
+#[cfg(unix)]
+pub(crate) struct FDReader {
```

Comment:
i'm not really well up on different options here in Rust tbh. Not sure what you mean by avoiding closing the fd. afaik, this method will just wrap the existing fd which is passed from the embedder side on every call. It shouldn't ever be closed on the rust side, unless there is something I am missing?

I just created this PR as a draft for discussion as I am not sure what the best approach would be in Rust but I think it would be nice to find an approach that avoided the callbacks which aren't strictly needed as they are used completely synchronously. 

This approach works exactly the same as the callback version for me in JS and I haven't found any issues in my testing so far. I can add some tests in C to this PR when i get some time if you think that would be useful.

Reply:
>  Not sure what you mean by avoiding closing the fd

I mean that if we used `File::from_raw_fd` to create a `File` from the passed in `fd` to read/write from instead of using this `FDReader` abstraction then without care when the `File` is dropped it would close the `fd`. I think we wouldn't want that to happen, and so would need to "leak" the `File` w/ `mem::forget` instead of letting it drop normally and document that the caller is responsible for closing the fd.